pkg/runtime/wfengine/activity.go

```diff
@@ -50,15 +72,68 @@ func (a *activityActor) SetActorRuntime(actorsRuntime actors.Actors) {
 // returns immediately after creating the reminder, enabling the workflow to continue processing other events
 // in parallel.
 func (a *activityActor) InvokeMethod(ctx context.Context, actorID string, methodName string, data []byte) (any, error) {
-	// This actor doesn't define any methods, and uses the method name and payload as the reminder name and payload.
-	err := a.createReliableReminder(ctx, actorID, methodName, data)
+	var ar ActivityRequest
+	if err := actors.DecodeInternalActorData(data, &ar); err != nil {
+		return nil, fmt.Errorf("failed to decode activity request: %w", err)
+	}
+
+	// Try to load activity state. If we find any, that means the activity invocation is a duplicate.
+	if state, err := a.loadActivityState(ctx, actorID, ar.Generation); err != nil {
+		return nil, err
+	} else if state.Generation > 0 {
+		return nil, ErrDuplicateInvocation
+	}
+
+	// Save the request details to the state store in case we need it after recovering from a failure.
+	state := activityState{
+		Generation:   ar.Generation,
+		EventPayload: ar.HistoryEvent,
+	}
+	if err := a.saveActivityState(ctx, actorID, state); err != nil {
+		return nil, err
+	}
+
+	// The actual execution is triggered by a reminder
+	err := a.createReliableReminder(ctx, actorID, "run-activity", ar.Generation)
 	return nil, err
 }
 
 // InvokeReminder implements actors.InternalActor and executes the activity logic.
-// TODO: Need to review how errors for reminders are handled.
 func (a *activityActor) InvokeReminder(ctx context.Context, actorID string, reminderName string, data any, dueTime string, period string) error {
-	taskEvent, err := backend.UnmarshalHistoryEvent(data.([]byte))
+	wfLogger.Debugf("invoking reminder '%s' on activity actor '%s'", reminderName, actorID)
+
+	generation := data.(uint64)
+	state, _ := a.loadActivityState(ctx, actorID, generation)
+	// TODO: On error, reply with a failure - this requires support from durabletask-go to produce TaskFailure results
+
+	timeoutCtx, cancelTimeout := context.WithTimeout(ctx, a.defaultTimeout)
+	defer cancelTimeout()
+
+	if err := a.executeActivity(timeoutCtx, actorID, reminderName, state.EventPayload); err != nil {
+		if err == timeoutCtx.Err() {
```

Comment:
Because `timeoutCtx` is a children of `ctx`, its error could also be `context.Canceled`.

I suggest changing this to be explicitly looking for `errors.Is(err, context.DeadlineExceeded)`

Reply:
(and maybe have a separate log for `context.Canceled`?)

Reply:
I changed this to use `errors.Is(err, context.DeadlineExceeded)` as you suggested. However, I didn't add a separate log for `context.Canceled` since we never attempt to explicitly cancel the context.